float32 tip_x
float32 tip_y
float32 tip_z